caproto/threading/client.py

```diff
@@ -300,6 +309,240 @@ def __call__(self):
+                entry.pop(addr)
+
+        if result_addr is not None:
+            return result_addr
+
+        raise CaprotoKeyError(f'{name!r}: stale search result')
+
+    @_locked
+    def received_search_response(self, cid, addr):
+        'Get an unanswered search by (cid, addr) -> name, queue'
+        first_response = True
+        try:
+            info = self._unanswered_searches.pop(cid)
+        except KeyError:
+            first_response = False
+            try:
+                info = self._searches[cid]
+            except KeyError:
+                # Completely unknown cid... ignore
+                raise UnknownSearchResponse('No matching cid') from None
+
+        name, queue, *_ = info
+        self.mark_name_found(name, addr)
+
+        if first_response:
+            return name, queue
+
+        raise DuplicateSearchResponse(name=name, cid=cid,
+                                      addresses=list(self.name_to_addrs[name]))
+
+    def items_to_retry(self, threshold, resend_deadline):
+        'All search results in need of retrying (if beyond threshold)'
+        with self._lock:
+            items = list(self._unanswered_searches.items())
+
+        if not threshold:
+            return items
+
+        # Skip over searches that haven't gotten any results in
+        # SEARCH_RETIREMENT_AGE.
+        return list((search_id, it)
+                    for search_id, it in items
+                    if (it[-1] > threshold and it[-2] < resend_deadline))
+
+    @_locked
+    def search(self, *names, results_queue, retirement_deadline):
+        'Search for names, adding items to results_queue'
+        # Search requests that are past their retirement deadline with no
+        # results will be searched for less frequently.
+        new_searches = dict(
+            (self._search_id_counter(),
+             [name, results_queue, 0, retirement_deadline]
+             )
+            for name in names)
+
+        self._unanswered_searches.update(new_searches)
+        self._searches.update(new_searches)
+
+        for id, info in new_searches.items():
+            name, *_ = info
+            self._searches_by_name[name] = info
+
+    @_locked
+    def check_cache(self, names):
+        """
+        Tell which PVs have valid cached addresses, and which do not.
+
+        Returns
+        -------
+        use_cached_search : dict
+            Address to list of names
+        needs_search : list
+            Remaining PVs that need a search attempt
+        """
+        needs_search = []
+        use_cached_search = defaultdict(list)
+
+        for name in names:
+            try:
+                address = self.get_cached_search_result(name)
+            except KeyError:
+                needs_search.append(name)
+            else:
+                use_cached_search[address].append(name)
+
+        return use_cached_search, needs_search
+
+
 class SharedBroadcaster:
     def __init__(self, *, registration_retry_time=10.0):
```

Comment:
Per your suggestion about moving to one-broadcaster-per-content, I guess `search_results` should an optional parameter to `Broadcaster.__init__`.

Reply:
I think so, but we may want to move that refactor to be done later. I'd like `SearchResults` to be done right (it still is not), and then move toward removing `SharedBroadcaster` separately.